### PR TITLE
deeply nested generics

### PR DESCRIPTION
**Pull request by wakemaster39**
@Tinche  Replaces #171. fixes the issue of `Optional[List[Sequence[Set[T]]]]` and `Optional[List[SomeGeneric]]` where the Generic does not have its parameters specified so mapping will be empty. 

* Adds default none type structuring.
* blacked the code base, not sure why it diverged but I can roll it back if you prefer.

### CODE REVIEW

**Comment by Tinche** [comment]
Cool! I blacked `bench/` on master in the meantime. Will take a look soon.

File: src/cattr/converters.py [hunk 1]
```diff
@@ -45,7 +45,7 @@
     make_hetero_tuple_unstructure_fn,
     make_iterable_unstructure_fn,
     make_mapping_structure_fn,
-    make_mapping_unstructure_fn,
+    make_mapping_unstructure_fn, _generate_mapping,
```
Comment (IceN9ne) [inline src/cattr/converters.py:48]: This should probably go on its own line to keep things regular.

**Comment by Tinche** [comment]
I can take a look at this in a week since I'm at a company retreat atm.

**Comment by wakemaster39** [comment]
Sounds good, I keep making some small tweaks and changes as I find more broken things around #170.

The crux of the issue is that unstructuring is based on the field type now and it is not based on the object being passed in. Deeply nested generics, protocols and others are basically hitting the `_unstructure_identity` and getting passed through.

As issues are found new handles will be needed. The possible recursion of the protocol handler that I created calling unstructure again has some smell to it, but I don't see another way right now to unstructure based on the type being passed in. 

This same methodology could also be used for TypeVar, so that there is a fallback handler of instead of hitting the identity. I have not run into it yet in my code but this seems like it could eventually run into issues with a TypeVar in the generic list:

https://github.com/Tinche/cattrs/blob/f377e04f9375d50c4496388f7b0f64d51790561c/src/cattr/gen.py#L292-L302

**Comment by petergaultney** [comment]
I can confirm that does indeed fix it. wow, that was fast. thank you!

**Comment by petergaultney** [comment]
I've done fairly extensive further testing and this seems to have resolved all of our issues with the new GenConverter in `typecats` and the applications it supports.

I would be very happy to see this merged whenever @Tinche and @wakemaster39 agree that it's ready.

**Comment by Tinche** [comment]
So while working on this the fact that `issubclass(int, Protocol) == True` is throwing me off a little. I was not expecting that, heh.

**Comment by Tinche** [comment]
My fix is live on main, including your tests! Let me know if it's good enough!

**Pull request closed without merging**